crates/nu-path/src/pwd_per_drive.rs

```diff
@@ -0,0 +1,449 @@
+/// use std::collections::HashMap;
+/// let mut env = HashMap::<String, String>::new();
+/// get_env_vars(&mut env);
+/// assert_eq!(env.get("=C:").unwrap(), r"C:\Users\Home");
+/// if let Some(expanded) = expand_pwd(Path::new("D:")) {
+///     let abs_path = expanded.as_path().to_str().expect("OK").to_string();
+///     if abs_path.len() > 3 {
+///         // ensure env var ends with '\\'
+///         let pwd_env = env.get("=D:").unwrap();
+///         let pwd_env_with_delimiter = if !pwd_env.ends_with('\\') && !pwd_env.ends_with('/') {
+///             format!(r"{}\", pwd_env)
+///         } else {
+///             pwd_env.to_string()
+///         };
+///         assert_eq!(pwd_env_with_delimiter, abs_path);
+///     }
+/// }
+/// ```
+use std::path::{Path, PathBuf};
+
+#[derive(PartialEq, Debug)]
+pub enum PathError {
+    InvalidDriveLetter,
+    InvalidPath,
+    CantLockSharedMap,
+}
+pub mod shared_map {
+    use super::*;
+    use crate::pwd_per_drive::PathError::CantLockSharedMap;
+
+    /// set_pwd_per_drive
+    /// Record PWD for drive, path must be absolute path
+    /// return Ok(()) if succeeded, otherwise error message
+    pub fn set_pwd(path: &Path) -> Result<(), PathError> {
+        if let Ok(mut pwd_per_drive) = get_shared_drive_pwd_map().lock() {
+            pwd_per_drive.set_pwd(path)
+        } else {
+            Err(CantLockSharedMap)
+        }
+    }
+
+    /// expand_pwe_per_drive
+    /// Input relative path, expand PWD-per-drive to construct absolute path
+    /// return PathBuf for absolute path, None if input path is invalid.
+    pub fn expand_pwd(path: &Path) -> Option<PathBuf> {
+        if need_expand(path) {
+            if let Ok(mut pwd_per_drive) = get_shared_drive_pwd_map().lock() {
+                return pwd_per_drive.expand_path(path);
+            }
+        }
+        None
+    }
+
+    /// Collect PWD-per-drive as env vars (for child process)
+    pub fn get_env_vars(env: &mut HashMap<String, String>) {
+        if let Ok(pwd_per_drive) = get_shared_drive_pwd_map().lock() {
+            pwd_per_drive.get_env_vars(env);
+        }
+    }
+}
+
+/// Helper to check if input path is relative path
+/// with drive letter, it can be expanded with PWD-per-drive.
+fn need_expand(path: &Path) -> bool {
+    if let Some(path_str) = path.to_str() {
+        let chars: Vec<char> = path_str.chars().collect();
+        if chars.len() >= 2 {
+            return chars[1] == ':' && (chars.len() == 2 || (chars[2] != '/' && chars[2] != '\\'));
+        }
+    }
+    false
+}
+
+struct DriveToPwdMap {
+    map: [Option<String>; 26], // Fixed-size array for A-Z
+}
+
+impl DriveToPwdMap {
+    pub fn new() -> Self {
+        // Initialize by current PWD-per-drive
+        let mut map: [Option<String>; 26] = Default::default();
+        for (drive_index, drive_letter) in ('A'..='Z').enumerate() {
+            let env_var = Self::env_var_for_drive(drive_letter);
+            if let Ok(env_pwd) = std::env::var(&env_var) {
```

Comment:
we typically don't allow `std::env::` since it's safer to use stack.get_env_var() and engine_state.get_env_var() and I don't think std::env::var is thread safe.

Reply:
OK, turn to use engine_state.get_env_var() instead.